src/InputAreaWidget.ts

```diff
@@ -0,0 +1,261 @@
+   */
+  updateMimetype(mimetype: string) {
+    if (CodeMirror.mimeModes.hasOwnProperty(mimetype)) {
+      this._editor.setOption('mode', mimetype);
+    } else {
+      let info = CodeMirror.findModeByMIME(mimetype);
+      if (info) {
+        this._loadCodeMirrorMode(info.mode).then(() => {
+          this._editor.setOption('mode', mimetype);
+        })
+      }
+    }
+  }
+
+  /**
+   * Update the line numbers in the editing widget.
+   */
+  updateLineNumbers(lineNumbers: boolean) {
+    this._editor.setOption('lineNumbers', lineNumbers);
+  }
+
+  /**
+   * Get the text in the widget.
+   */
+  get text(): string {
+    return this._editor.getDoc().getValue();
+  }
+
+  /**
+   * Set the text in the widget.
+   */
+  set text(text: string) {
+    this.updateText(text)
+  }
+
+  /**
+   * Handle afterAttach messages.
+   */
+  protected onAfterAttach(msg: Message): void {
+    this._editor.refresh();
+  }
+
+  /**
+   * Handle resize messages.
+   */
+  protected onResize(msg: ResizeMessage): void {
+    if (msg.width < 0 || msg.height < 0) {
+      this._editor.refresh();
+    } else {
+      this._editor.setSize(msg.width, msg.height);
+    }
+  }
+
+  /**
+   * Change handler for model updates.
+   */
+  private _modelUpdate(sender: ITextEditorViewModel, args: IChangedArgs<any>) {
+    switch(args.name) {
+    case 'fixedHeight':
+      this.updateFixedHeight(args.newValue);
+      break;
+    case 'text':
+      this.updateText(args.newValue);
+      break;
+    case 'mimetype':
+      this.updateMimetype(args.newValue);
+      break;
+    case 'lineNumbers':
+      this.updateLineNumbers(args.newValue);
+      break;
+    }
+  }
+
+  /**
+   * Load a CodeMirror mode asynchronously.
+   */
+  private _loadCodeMirrorMode(mode: string) {
+    // load codemirror mode module, returning a promise.
+    if (CodeMirror.modes.hasOwnProperty(mode)) {
+      return Promise.resolve();
+    } else {
+      return System.import(`codemirror/mode/${mode}/${mode}`);
+    }
+  }
+
+  private _editor: CodeMirror.Editor;
+  private _model: ITextEditorViewModel;
+}
+
+export
+interface FunctionTable {
+  [key: string]: Function
+}
+
+/**
+ * An input area widget, which hosts an editor widget.
+ */
+export
+class InputAreaWidget extends Panel {
+
+  static update: FunctionTable = {
+    'textEditor': InputAreaWidget.prototype.updateTextEditor,
+    'rendered': InputAreaWidget.prototype.updateRendered,
+    'collapsed': InputAreaWidget.prototype.updateCollapsed,
+    'promptNumber': InputAreaWidget.prototype.updatePromptNumber,
+    'executionCount': InputAreaWidget.prototype.updateExecutionCount
+  }
+
+  /**
+   * Construct an input area widget.
+   */
+  constructor(model: IInputAreaViewModel) {
+    super();
+    this.addClass('InputAreaWidget');
+    this._model = model;
+    for (let key in InputAreaWidget.update) {
+      InputAreaWidget.update[key].call(this, model[key]);
```

Comment:
yep, this was one of the experiments, and I already decided to scrap it.  I have some commits deleting it that I haven't pushed.